sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -127,6 +127,165 @@ case class MapKeys(child: Expression)
   override def prettyName: String = "map_keys"
 }
 
+@ExpressionDescription(
+  usage = """_FUNC_(a1, a2, ...) - Returns a merged array containing in the N-th position the
+  N-th value of each array given.""",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3), array(2, 3, 4));
+        [[1, 2], [2, 3], [3, 4]]
+      > SELECT _FUNC_(array(1, 2), array(2, 3), array(3, 4));
+        [[1, 2, 3], [2, 3, 4]]
+  """,
+  since = "2.4.0")
+case class Zip(children: Seq[Expression]) extends Expression with ExpectsInputTypes {
+
+  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.length)(ArrayType)
+
+  override def dataType: DataType = ArrayType(mountSchema)
+
+  override def nullable: Boolean = children.forall(_.nullable)
+
+  private lazy val arrayTypes = children.map(_.dataType.asInstanceOf[ArrayType])
+
+  private lazy val arrayElementTypes = arrayTypes.map(_.elementType)
+
+  def mountSchema: StructType = {
+    val fields = children.zip(arrayElementTypes).zipWithIndex.map {
+      case ((expr: NamedExpression, elementType), _) =>
+        StructField(expr.name, elementType, nullable = true)
+      case ((_, elementType), idx) =>
+        StructField(s"$idx", elementType, nullable = true)
+    }
+    StructType(fields)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val numberOfArrays: Int = children.length
+    val genericArrayData = classOf[GenericArrayData].getName
+    val genericInternalRow = classOf[GenericInternalRow].getName
+    val arrVals = ctx.freshName("arrVals")
+    val arrCardinality = ctx.freshName("arrCardinality")
+    val biggestCardinality = ctx.freshName("biggestCardinality")
+    val storedArrTypes = ctx.freshName("storedArrTypes")
+    val returnNull = ctx.freshName("returnNull")
+    val evals = children.map(_.genCode(ctx))
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+        |${eval.code}
+        |if (!${eval.isNull}) {
+        |  $arrVals[$index] = ${eval.value};
+        |  $arrCardinality[$index] = ${eval.value}.numElements();
+        |} else {
+        |  $arrVals[$index] = null;
+        |  $arrCardinality[$index] = 0;
+        |  $returnNull[0] = true;
+        |}
+        |$storedArrTypes[$index] = "${arrayElementTypes(index)}";
```

Comment:
Do we need `storedArrType`? Since we can know data type of all children before execution, would it be possible to check the correctness in own `checkInputDataTypes()`?

Reply:
I'm not using `storedArrType` to check if the type is correct, but rather to be able to pick the correct `getValue` for the current type.

Reply:
In simple cases, since we know only a data type of all children before execution, we may not need to use `$storedArrTypes`. However, I may miss something. 

Would it be possible to show an example test case that requires to pick the correct `getValue` by using `$storedArrTypes`?

Reply:
Suppose that you have three inputs, where the first is an Array[Int], the second an Array[String], and third an Array[Long]. How do you, in runtime, uses `getInt(s"$arrVals[$j]", IntegerType, i)` only when filling **the first position of each row**, `getString(s"$arrVals[$j]", StringType, i)` only when filling the second position of each row and `getInt(s"$arrVals[$j]", LongType, i)` only when filling the third? 
My solution is: I hardcodded a condition: if the stored arrayType for the current array is "StringType", when I use `getValue(..., StringType, i)`, if it is "DoubleType", then I use `getValue(..., DoubleType, i)`, etc.